### ISSUE TITLE
[🐛 Bug]: UnreachableBrowserException caused by "connection closed locally" exactly 3 minutes after session start when using JDK HttpClient

### ISSUE THREAD

**Issue by robbytx** [SeleniumHQ/selenium#11798]
### What happened?

I recently started using `selenium-http-jdk-client` as a workaround for https://github.com/SeleniumHQ/selenium/issues/11750#issuecomment-1460580132, and I've noticed that tests which run longer than 3 minutes often fail with the following exception:
```
Exception in thread "main" org.openqa.selenium.remote.UnreachableBrowserException: Error communicating with the remote browser. It may have died.
Build info: version: '4.8.1', revision: '8ebccac989'
System info: os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '12.6.3', java.version: '11.0.15'
Driver info: org.openqa.selenium.remote.RemoteWebDriver
Command: [60c7ca2458c8472f824c2c43dbf1ac2a, sendKeysToElement {id=ca596158-ff3e-4df8-b4ac-2c00e60ca6d7, value=[Ljava.lang.CharSequence;@474e34e4}]
Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 111.0.5563.65, chrome: {chromedriverVersion: 111.0.5563.41 (976ef12907ef..., userDataDir: C:\Program Files (x86)\scop...}, goog:chromeOptions: {debuggerAddress: localhost:49762}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: WINDOWS, proxy: Proxy(), setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:extension:minPinLength: true, webauthn:extension:prf: true, webauthn:virtualAuthenticators: true}
Element: [[RemoteWebDriver: chrome on WINDOWS (60c7ca2458c8472f824c2c43dbf1ac2a)] -> css selector: input[name='q']]
Session ID: 60c7ca2458c8472f824c2c43dbf1ac2a
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:567)
	at org.openqa.selenium.remote.RemoteWebElement.execute(RemoteWebElement.java:257)
	at org.openqa.selenium.remote.RemoteWebElement.sendKeys(RemoteWebElement.java:113)
	at example.Example.main(Example.java:42)
Caused by: java.io.UncheckedIOException: java.io.IOException: HTTP/1.1 header parser received no bytes
	at org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute(JdkHttpClient.java:296)
	at org.openqa.selenium.remote.tracing.TracedHttpClient.execute(TracedHttpClient.java:55)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:181)
	at org.openqa.selenium.remote.TracedCommandExecutor.execute(TracedCommandExecutor.java:51)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:543)
	... 3 more
Caused by: java.io.IOException: HTTP/1.1 header parser received no bytes
	at java.net.http/jdk.internal.net.http.HttpClientImpl.send(HttpClientImpl.java:565)
	at java.net.http/jdk.internal.net.http.HttpClientFacade.send(HttpClientFacade.java:119)
	at org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute(JdkHttpClient.java:292)
	... 7 more
Caused by: java.io.IOException: HTTP/1.1 header parser received no bytes
	at java.net.http/jdk.internal.net.http.common.Utils.wrapWithExtraDetail(Utils.java:327)
	at java.net.http/jdk.internal.net.http.Http1Response$HeadersReader.onReadError(Http1Response.java:673)
	at java.net.http/jdk.internal.net.http.Http1AsyncReceiver.checkForErrors(Http1AsyncReceiver.java:297)
	at java.net.http/jdk.internal.net.http.Http1AsyncReceiver.flush(Http1AsyncReceiver.java:263)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$SynchronizedRestartableTask.run(SequentialScheduler.java:175)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:147)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:198)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.io.IOException: connection closed locally
	at java.net.http/jdk.internal.net.http.SocketTube.signalClosed(SocketTube.java:157)
	at java.net.http/jdk.internal.net.http.PlainHttpConnection.close(PlainHttpConnection.java:277)
	at java.net.http/jdk.internal.net.http.AsyncSSLConnection.close(AsyncSSLConnection.java:111)
	at java.net.http/jdk.internal.net.http.Http1Exchange.cancelImpl(Http1Exchange.java:490)
	at java.net.http/jdk.internal.net.http.Http1Exchange.cancel(Http1Exchange.java:426)
	at java.net.http/jdk.internal.net.http.Exchange.cancel(Exchange.java:228)
	at java.net.http/jdk.internal.net.http.MultiExchange.cancel(MultiExchange.java:240)
	at java.net.http/jdk.internal.net.http.ResponseTimerEvent.handle(ResponseTimerEvent.java:71)
	at java.net.http/jdk.internal.net.http.HttpClientImpl.purgeTimeoutsAndReturnNextDeadline(HttpClientImpl.java:1248)
	at java.net.http/jdk.internal.net.http.HttpClientImpl$SelectorManager.run(HttpClientImpl.java:877)
```

My analysis of the stack trace indicates that the JDK HttpClient is firing a `ResponseTimerEvent` exactly 3 minutes after the `newSession` request was made, even though the request appears to have successfully completed, and requests are subsequently being made without issue.  The 3-minute duration appears to stem from the default `ClientConfig.readTimeout`, and when it elapses, the request that is currently awaiting a response (if any) appears to become the victim of the spurious timeout.

From my understanding, this feels like a bug in the JDK client -- I'm not seeing any misbehavior on Selenium's part when I step through the logic -- but I've not seen any bug report related to this, so I'm starting here seeing as the `selenium-http-jdk-client` is relatively newer and lightly used compared to the JDK.

### How can we reproduce the issue?

```shell
package example;

import org.openqa.selenium.By;
import org.openqa.selenium.Platform;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.remote.Browser;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.openqa.selenium.remote.RemoteWebDriver;

import java.net.URI;
import java.net.URL;
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;

public class Example {

    public static void main(String[] args) throws Exception {
        System.setProperty("webdriver.http.factory", "jdk-http-client"); // issue only occurs with the jdk client?

        DesiredCapabilities browserOptions = new DesiredCapabilities();
        browserOptions.setBrowserName(Browser.CHROME.browserName());
        browserOptions.setPlatform(Platform.WIN10);
        Map<String, Object> sauceOptions = new HashMap<>();
        sauceOptions.put("name", "selenium-jdk-http-timeout");
        browserOptions.setCapability("sauce:options", sauceOptions);

        URL url = new URI("https",
                System.getProperty("sauce.userinfo"),
                "ondemand.us-west-1.saucelabs.com", 443, "/wd/hub",
                null, null
        ).toURL();

        Instant start = Instant.now();
        RemoteWebDriver driver = new RemoteWebDriver(url, browserOptions);
        driver.navigate().to("https://www.google.com");

        Thread.sleep(5000);
        for (Instant now = Instant.now(); now.isBefore(start.plusSeconds(200)); now = Instant.now()) {
            WebElement query = driver.findElement(By.cssSelector("input[name='q']"));
            query.clear();
            query.sendKeys(Instant.now().toString());
//            Thread.sleep(5000); // uncommenting this sleep appears to prevent the problem (since nothing is waiting when the timer expires?)
        }
    }
}
```

Reproduced with `org.seleniumhq.selenium:selenium-remote-driver:4.8.1` and `org.seleniumhq.selenium:selenium-http-jdk-client:4.8.1` as Maven dependencies, using the following JDK:
```
openjdk version "11.0.15" 2022-04-19
OpenJDK Runtime Environment Temurin-11.0.15+10 (build 11.0.15+10)
OpenJDK 64-Bit Server VM Temurin-11.0.15+10 (build 11.0.15+10, mixed mode)
```

### Relevant log output

See stack trace above.

### Operating System

macOS Monterey

### Selenium version

OpenJDK 64-Bit Server VM Temurin-11.0.15+10

### What are the browser(s) and version(s) where you see this issue?

Chrome 111 on Windows 10

### What are the browser driver(s) and version(s) where you see this issue?

RemoteWebDriver 4.8.1

### Are you using Selenium Grid?

SauceLabs

**Comment by robbytx** [comment]
After further investigation, I'm able to reproduce the issue (or at least one with a similar cause) directly using the JDK HttpClient (without Selenium involved):
```java
package example;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.time.Instant;

import static java.net.http.HttpClient.Redirect.ALWAYS;
import static java.net.http.HttpClient.Version.HTTP_1_1;

public class JavaExample {

    public static void main(String[] args) throws Exception {
        int timeoutSeconds = 10;

        HttpClient client = HttpClient.newBuilder()
                .followRedirects(ALWAYS)
                .build();

        for (Instant start = Instant.now(), now = start; now.isBefore(start.plusSeconds(timeoutSeconds * 3 / 2)); now = Instant.now()) {
            HttpRequest request = HttpRequest.newBuilder().uri(URI.create("https://github.com/SeleniumHQ/selenium/releases/download/selenium-4.8.0/selenium-java-4.8.1.zip"))
                    .GET()
                    .version(HTTP_1_1)
                    .timeout(Duration.ofSeconds(timeoutSeconds))
                    .build();
            HttpResponse<Void> response = client.send(request, HttpResponse.BodyHandlers.discarding());
            System.out.println(now + " -> " + response.headers().firstValue("date"));
        }
    }
}
```

After ~10 seconds will typically result in an exception like:
```
Exception in thread "main" java.io.IOException: fixed content-length: 28490775, bytes received: 16613376
	at java.net.http/jdk.internal.net.http.HttpClientImpl.send(HttpClientImpl.java:565)
	at java.net.http/jdk.internal.net.http.HttpClientFacade.send(HttpClientFacade.java:119)
	at example.JavaExample.main(JavaExample.java:30)
Caused by: java.io.IOException: fixed content-length: 28490775, bytes received: 16613376
	at java.net.http/jdk.internal.net.http.common.Utils.wrapWithExtraDetail(Utils.java:330)
	at java.net.http/jdk.internal.net.http.Http1Response$BodyReader.onReadError(Http1Response.java:758)
	at java.net.http/jdk.internal.net.http.Http1AsyncReceiver.checkForErrors(Http1AsyncReceiver.java:297)
	at java.net.http/jdk.internal.net.http.Http1AsyncReceiver.flush(Http1AsyncReceiver.java:263)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$SynchronizedRestartableTask.run(SequentialScheduler.java:175)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:147)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:198)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.io.IOException: connection closed locally
	at java.net.http/jdk.internal.net.http.SocketTube.signalClosed(SocketTube.java:157)
	at java.net.http/jdk.internal.net.http.PlainHttpConnection.close(PlainHttpConnection.java:277)
	at java.net.http/jdk.internal.net.http.AsyncSSLConnection.close(AsyncSSLConnection.java:111)
	at java.net.http/jdk.internal.net.http.Http1Exchange.cancelImpl(Http1Exchange.java:490)
	at java.net.http/jdk.internal.net.http.Http1Exchange.cancel(Http1Exchange.java:426)
	at java.net.http/jdk.internal.net.http.Exchange.cancel(Exchange.java:228)
	at java.net.http/jdk.internal.net.http.MultiExchange.cancel(MultiExchange.java:240)
	at java.net.http/jdk.internal.net.http.ResponseTimerEvent.handle(ResponseTimerEvent.java:71)
	at java.net.http/jdk.internal.net.http.HttpClientImpl.purgeTimeoutsAndReturnNextDeadline(HttpClientImpl.java:1248)
	at java.net.http/jdk.internal.net.http.HttpClientImpl$SelectorManager.run(HttpClientImpl.java:836)
```

I get the same behavior on all the following JDKs:
- OpenJDK 64-Bit Server VM Temurin-11.0.18+10
- OpenJDK 64-Bit Server VM Temurin-17.0.6+10
- OpenJDK 64-Bit Server VM (build 20+36-2344) (released today!)

So I'll file a JDK issue instead and link it here.

**Comment by robbytx** [comment]
Reported as https://bugs.java.com/bugdatabase/view_bug?bug_id=8304701 / https://bugs.openjdk.org/browse/JDK-8304701

**Comment by robbytx** [comment]
AFAICT this issue will affect any test:

- Run on Java 11+
- Using the Java HTTP client (`selenium-jdk-http-client`)
- Where RemoteWebDriver connects to a server
- Server responds with a HTTP/1.1 (persistent) connection
- Test runs for 3+ minutes actively (i.e. there is frequently a Selenium request in-flight)

Heads up to @titusfortner, since you wrote https://github.com/SeleniumHQ/selenium/issues/11750#issuecomment-1460557716:

> AsyncHttp Client has not been actively maintained (looks like this just changed), so we've been requesting people upgrade to the Java 11 HTTP client https://www.selenium.dev/blog/2022/using-java11-httpclient/

While that issue concerned local ChromeDriver and this one concerns RemoteWebDriver, this issue leads me to question -- or at least caveat -- the recommendation for people to upgrade to the Java 11 HTTP client.

**Comment by joerg1985** [comment]
This seems to be related to the redirect, without the redirect it will work.
Therefore i think the HttpClient implementation of selenium could handle the redirect and the issue should be fixed.

```
 public static void main(String[] args) throws Exception {
        int timeoutSeconds = 10;

        HttpClient client = HttpClient.newBuilder()
                .followRedirects(HttpClient.Redirect.NEVER)
                .build();

        for (Instant start = Instant.now(), now = start; now.isBefore(start.plusSeconds(timeoutSeconds * 3 / 2)); now = Instant.now()) {
            HttpRequest request = HttpRequest.newBuilder().uri(URI.create("https://github.com/SeleniumHQ/selenium/releases/download/selenium-4.8.0/selenium-java-4.8.1.zip"))
                    .GET()
                    .version(HTTP_1_1)
                    .timeout(Duration.ofSeconds(timeoutSeconds))
                    .build();
                            
            HttpResponse<Void> response = client.send(request, HttpResponse.BodyHandlers.discarding());
            
            if (response.statusCode() == 302) {
                 URI location = URI.create(response.headers().firstValue("location").get());
                 
                 request = HttpRequest.newBuilder().uri(location)
                    .GET()
                    .version(HTTP_1_1)
                    .timeout(Duration.ofSeconds(timeoutSeconds))
                    .build();
                            
                 response = client.send(request, HttpResponse.BodyHandlers.discarding());
            }
            
            System.out.println(now + " -> " + response.headers().firstValue("date"));
        }
    }
```

**Comment by robbytx** [comment]
Indeed, you are correct.  I've submitted an update to the JDK bug report to clarify this point.

**Comment by titusfortner** [comment]
So current code always fails if initial session request receives a redirect?

We need to be able to support redirects, though, so I'm not sure it's a good fix to just turn that off?

**Comment by joerg1985** [comment]
@titusfortner The workaround would not be to disable redirects, it is about handling the redirects inside the selenium http client and not inside the jdk client.

**Comment by titusfortner** [comment]
Thanks for clarification! 
Can you make a PR? 😄

**Comment by joerg1985** [comment]
@titusfortner sure, might take some days. But i think this is not time critical.

**Comment by diemol** [comment]
Should be fixed by #11816, thanks @joerg1985!

The next release should happen in ~1 week.

**Comment by quakenstein** [comment]
Is this issue really fixed? With Selenium 4.14.1 i get excatly this erros running tests with remotewebdriver to browserstack,

**Comment by rajadinesh94** [comment]
PY/BOM/0010088/000/323900- pf account number .
